File: examples/python/gateway_chat.py
from __future__ import annotations

from sipp import (
    CacheRuntimeConfig,
    ChatMessage,
    SippClient,
    SippTextOptions,
    SippTextRun,
    ContextRuntimeConfig,
    GatewayDescriptor,
    LocalModelDescriptor,
    LocalTextOptions,
    ModelPlacementConfig,
    NativeRuntimeConfig,
    ObservabilityRuntimeConfig,
    ResidencyRuntimeConfig,
    SamplingRuntimeConfig,
    SchedulerRuntimeConfig,
    set_llama_log_quiet,
)

from _support import (
    DEFAULT_CONTEXT,
    DEFAULT_MAX_TOKENS,
    DEFAULT_SEED,
    DEFAULT_TEMPERATURE,
    DEFAULT_TOP_P,
    float_env,
    gpu_layers,
    int_env,
    print_text,
    read_gateway_args,
    required_env,
)


def runtime_config(*, embeddings: bool) -> NativeRuntimeConfig:
    return NativeRuntimeConfig(
        placement=ModelPlacementConfig(gpu_layers=gpu_layers()),
        context=ContextRuntimeConfig(
            n_ctx=int_env("SIPP_CONTEXT", DEFAULT_CONTEXT),
            n_threads=int_env("SIPP_THREADS"),
            n_threads_batch=int_env("SIPP_THREADS"),
            embeddings=embeddings,
        ),
        sampling=SamplingRuntimeConfig(
            temperature=float_env("SIPP_TEMPERATURE", DEFAULT_TEMPERATURE),
            seed=int_env("SIPP_SEED", DEFAULT_SEED),
        ),
        scheduler=SchedulerRuntimeConfig(
            continuous_batching=True,
            prefill_chunk_size=0,
        ),
        cache=CacheRuntimeConfig(mode="live_slot_prefix"),
        residency=ResidencyRuntimeConfig(max_gpu_models_per_device=1),
        observability=ObservabilityRuntimeConfig(runtime_metrics=True),
    )


def text_options() -> SippTextOptions:
    return SippTextOptions(
        max_tokens=int_env("SIPP_MAX_TOKENS", DEFAULT_MAX_TOKENS),
        temperature=float_env("SIPP_TEMPERATURE", DEFAULT_TEMPERATURE),
        top_p=float_env("SIPP_TOP_P", DEFAULT_TOP_P),
    )


def chat_messages(prompt: str) -> list[ChatMessage]:
    return [
        ChatMessage("system", "Answer concisely."),
        ChatMessage("user", prompt),
    ]


def collect_streamed_text(label: str, run: SippTextRun) -> dict[str, object]:
    streamed = ""
    print(f"{label}_stream=", end="", flush=True)
    for batch in run.tokens():
        print(batch["text"], end="", flush=True)
        streamed += batch["text"]
    print()
    result = run.result()
    if streamed != result["text"]:
        raise RuntimeError("streamed token batches did not match final response text")
    return result


def main() -> None:
    model, target, prompt = read_gateway_args(
        "gateway_chat", "Explain gateway-backed inference in one sentence."
    )
    set_llama_log_quiet(True)

    client = SippClient()
    local_endpoint = client.add(
        "local",
        LocalModelDescriptor(model, runtime_config(embeddings=False)),
    )
    gateway_endpoint = client.add(
        "gateway",
        GatewayDescriptor(
            target,
            required_env("SIPP_GATEWAY_URL"),
            authentication_kind="bearer",
            authentication_value=required_env("SIPP_GATEWAY_TOKEN"),
        )
    )

    local_run = client.chat(
        chat_messages(prompt),
        endpoint=local_endpoint,
        options=text_options(),
        local=LocalTextOptions(context_key="python-gateway-chat-local"),
        emit_tokens=True,
    )
    local = collect_streamed_text("local", local_run)

    gateway_run = client.chat(
        chat_messages(prompt),
        endpoint=gateway_endpoint,
        options=text_options(),
        emit_tokens=True,
    )
    gateway = collect_streamed_text("gateway", gateway_run)

    print("local:")
    print_text(local)
    print("gateway:")
    print_text(gateway)


if __name__ == "__main__":
    main()

File: examples/rust/src/gateway_chat.rs
mod support;

use std::path::PathBuf;

use futures::executor::block_on;
use futures::StreamExt;
use sipp::backend::set_llama_log_quiet;
use sipp::engine::{
    CacheRuntimeConfig, ContextRuntimeConfig, GpuLayerConfig, KvReuseMode, ModelPlacementConfig,
    NativeRuntimeConfig, ObservabilityRuntimeConfig, ResidencyRuntimeConfig, SamplingRuntimeConfig,
    SchedulerRuntimeConfig,
};
use sipp::engine::{ChatMessage, ChatRole};
use sipp::{
    EndpointDescriptor, GatewayAuthentication, GatewayEndpointConfig, GatewayRoutes, GatewaySecret,
    GatewayTimeoutPolicy, LocalTextOptions, SippChatRequest, SippClient, SippTextOptions,
    SippTextResponse, SippTextRun,
};

fn main() -> support::ExampleResult<()> {
    block_on(async {
        let args = support::gateway_args(
            "Explain gateway-backed inference in one sentence.",
            "gateway_chat",
        )?;
        set_llama_log_quiet(true);

        let mut client = SippClient::new();
        let local_endpoint = client
            .add(
                "local",
                EndpointDescriptor::local(args.model_path, runtime_config(false, None)),
            )
            .await?;
        let config = GatewayEndpointConfig {
            target: args.target.clone(),
            base_url: support::required_env("SIPP_GATEWAY_URL")?,
            routes: GatewayRoutes::default(),
            authentication: GatewayAuthentication::Bearer(GatewaySecret::new(
                support::required_env("SIPP_GATEWAY_TOKEN")?,
            )),
            static_headers: Default::default(),
            timeouts: GatewayTimeoutPolicy::default(),
            protocol_options: Default::default(),
        };
        let gateway_endpoint = client
            .add("gateway", EndpointDescriptor::gateway(config))
            .await?;

        let local_run = client.chat(SippChatRequest {
            endpoint: Some(local_endpoint),
            messages: chat_messages(args.input.clone()),
            options: text_options(),
            local: LocalTextOptions {
                context_key: Some("rust-gateway-chat-local".to_string()),
                ..Default::default()
            },
            emit_tokens: true,
            ..Default::default()
        });
        let local = collect_streamed_text("local", local_run).await?;

        let gateway_run = client.chat(SippChatRequest {
            endpoint: Some(gateway_endpoint),
            messages: chat_messages(args.input),
            options: text_options(),
            emit_tokens: true,
            ..Default::default()
        });
        let gateway = collect_streamed_text("gateway", gateway_run).await?;

        println!("local:");
        support::print_text(local);
        println!("gateway:");
        support::print_text(gateway);
        Ok(())
    })
}

fn chat_messages(input: String) -> Vec<ChatMessage> {
    vec![
        ChatMessage::new(ChatRole::System, "Answer concisely."),
        ChatMessage::new(ChatRole::User, input),
    ]
}

async fn collect_streamed_text(
    label: &str,
    run: SippTextRun,
) -> support::ExampleResult<SippTextResponse> {
    let (mut tokens, response) = run.into_parts();
    let mut streamed = String::new();
    print!("{label}_stream=");
    while let Some(batch) = tokens.next().await {
        print!("{}", batch.text);
        streamed.push_str(&batch.text);
    }
    println!();

    let response = response.await?;
    if streamed != response.text {
        return Err("streamed token batches did not match final response text".into());
    }
    Ok(response)
}

fn runtime_config(embeddings: bool, projector_path: Option<PathBuf>) -> NativeRuntimeConfig {
    NativeRuntimeConfig {
        placement: ModelPlacementConfig {
            gpu_layers: support::env_parse("SIPP_GPU_LAYERS")
                .map(GpuLayerConfig::from_layer_count)
                .unwrap_or(GpuLayerConfig::Auto),
            ..Default::default()
        },
        context: ContextRuntimeConfig {
            n_ctx: support::env_parse("SIPP_CONTEXT").or(Some(support::DEFAULT_CONTEXT)),
            n_threads: support::env_parse("SIPP_THREADS"),
            n_threads_batch: support::env_parse("SIPP_THREADS"),
            embeddings: embeddings.then_some(true),
            ..Default::default()
        },
        sampling: SamplingRuntimeConfig {
            temperature: support::env_parse("SIPP_TEMPERATURE")
                .or(Some(support::DEFAULT_TEMPERATURE)),
            seed: support::env_parse("SIPP_SEED").or(Some(support::DEFAULT_SEED)),
            ..Default::default()
        },
        scheduler: SchedulerRuntimeConfig {
            continuous_batching: true,
            prefill_chunk_size: 0,
            ..Default::default()
        },
        cache: CacheRuntimeConfig {
            mode: KvReuseMode::LiveSlotPrefix,
            ..Default::default()
        },
        multimodal: sipp::engine::MultimodalRuntimeConfig {
            projector_path: projector_path.map(|path| path.to_string_lossy().into_owned()),
            ..Default::default()
        },
        residency: ResidencyRuntimeConfig {
            max_gpu_models_per_device: 1,
            ..Default::default()
        },
        observability: ObservabilityRuntimeConfig {
            runtime_metrics: true,
            backend_profiling: false,
        },
    }
}

fn text_options() -> SippTextOptions {
    SippTextOptions {
        max_tokens: support::env_parse("SIPP_MAX_TOKENS").or(Some(support::DEFAULT_MAX_TOKENS)),
        temperature: support::env_parse("SIPP_TEMPERATURE").or(Some(support::DEFAULT_TEMPERATURE)),
        top_p: support::env_parse("SIPP_TOP_P").or(Some(support::DEFAULT_TOP_P)),
        stop: Vec::new(),
    }
}
